### DISCUSSION TITLE
Camera Control

### DISCUSSION THREAD

**Discussion by HR1959** [FactoryXCode/MfPack#76] [Q&A] [answered]
Dear all,

I try to use MfPack for a little apllication with a Web Cam and would like to control some camera function like autofocus on / off and focus setting.

I wanted to use   IID_IAMCameraControl   : TGUID = '{C6E13370-30AC-11d0-A18C-00A0C9118956}'; because I found with DirectShowCaptureCapabilities-x64.exe from https://alax.info/blog/1531 that my camera supports this interface.

But with QueryInterface I get allsways the result $80004002 means  E_NOINTERFACE,  no such interface supported.

Is there a function in your source code what allows to use the interface CameraControl where I can learn out of it how to use it properly?

Regards

Harals Redelberger

**Comment by FactoryXCode** [comment]
Hello,  Harals.

IAMCameraControl is a DirectShow interface.
But this is not going to help you when using Media Foundation.

To expose the Media Foundation interfaces it's essential to install the correct camera driver (Windows Driver Model (WDM) ) for Windows 10 and 11.  [See here.](https://learn.microsoft.com/en-us/windows/win32/api/strmif/nn-strmif-iamcameracontrol)

MfPack has a sample SimpleVideoCapture [see here](https://github.com/FactoryXCode/MfPack/tree/Master/MfPack/Samples/MfSimpleCapture)  to show you the basics about capturing from a camera. This example shows you how to use IMFMediaSource and IMFVideoDisplayControl. Remember these are low-level interfaces, but also the most flexible ones.

Other options like the IMFCameraControlDefaults interface gives you more control over your camera [see here](https://learn.microsoft.com/en-us/windows/win32/api/mfidl/nn-mfidl-imfcameracontroldefaults)
There is an example how to retrieve the default camera settings and how to change and store them.
If you have problems to translate the C++ sample, let me know.

Here is a sample code to accomplish on what -I think- you want:
I tested this roughly in the MfSimpleCapture sample, and works as expected.

```pascal
// Add WinApi.StrmIf to uses clause.
function TMfCaptureEngine.GetCameraSettings(): HRESULT;
var
  hr: HRESULT;
  mfSource: IMFMediaSource;
  amCameraControl: IAMCameraControl;
  amVidProcAmp: IAMVideoProcAmp;
  rMin: LONG;
  rMax: LONG;
  rDelta: LONG;
  rDeflt: LONG;
  cflag: CameraControlFlags;
  vflag: VideoProcAmpFlags;

begin

  mfSource := m_pSource; // Defined in private section.

  if (mfSource <> nil) then
    begin
      try
        // Control the 'camera'.
        amCameraControl := mfSource as IAMCameraControl;
        if Assigned(amCameraControl) then
          begin

            hr := amCameraControl.GetRange(LONG(CameraControl_Exposure),
                                           rMin,
                                           rMax,
                                           rDelta,
                                           rDeflt,
                                           LONG(cflag));

            // Add to public section of your class.
            // MfDeviceCapture.CameraController := amCameraControl;
          end;

        // Control the 'video'.
        amVidProcAmp := mfSource as IAMVideoProcAmp;
        if Assigned(amVidProcAmp) then
          begin
            hr := amVidProcAmp.GetRange(VideoProcAmp_Contrast,
                                        rMin,
                                        rMax,
                                        rDelta,
                                        rDeflt,
                                        LONG(vflag));
            // Add to public section of your class.
            // MfDeviceCapture.VideoController := amVidProcAmp;
          end;

      except
        on E: Exception do
          begin
            // Let user know an error has been occured.
            // ShowErrorMessage();
          end;
      end;
    end;
end;

``` 

Maybe this information will help you further.
Regards Tony.

**Comment by FactoryXCode** [comment, marked as answer]
Ok, I created [MfSimpleCapture Sample 2](https://github.com/FactoryXCode/MfPack/tree/Master/MfPack/Samples/MfSimpleCapture/Sample%202), where both interfaces are implemented. 
Please note that depending on your camera drivers or capabillities some settings could be unsupported.

I hope this will help you further on, 
Regards Tony.

**Reply by HR1959** [reply]
Dear Tony,

thank you very much for your effort in supporting me.
It is EXACTLY what I am looking for!
I downloaded immediately your second sample.
But unfortunately, when I compile it, I get 12 type conflicts: tagVideoProcAmpProperty versus Integer.

Because I am not familiar with all what you have done and it concerns only the video properties (I am mainly interested in the camera properties first) I just commented the lines raising the errors and compiled it again with success obviously.

But when I select a camara I get no picture on none of my computers.

To be sure that my cameras do not generate the issue I compiled the first sample MfSimpleCapture from the MfPack317 and it works fine on all my computers, means I get the video on the screen.

Can you please give a look again. I am sure you will find the issues much, much faster than me.

Thank you in advance.

Bets regards
Harald Redelberger
by the way, in my first message there is a typo in my first name … I am getting old :)

**Reply by FactoryXCode** [reply]
Hi Harald.

I've recently updated MfSimpleCapture Sample 2, WinApi.MediaFoundationApi.MfMetLib and WinApi.StrmIf to solve the conflicts you encountered.

Let me know if this solved your issues.
I tested the sample with Delphi 10.4.

Regards Tony.

**Reply by HR1959** [reply]
Dear Tony,

this time I downloaded the full MfPack as a zip file.
I deinstalled the old MfPack and installed the new version including the updated files you mentioned.
Then I compiled MfSimpleCapture Sample 2.
Unfortunately I get an error message when I select the camera:
![Screenshot 2024-09-28 102208](https://github.com/user-attachments/assets/e8cd823f-579a-4d04-9169-668ae1fabfe2)
It says that the interface is not supported.
I tested also the other sample MfSimpleCapture and get the same result.
Then I went back to the previous version of MfPack and surprisingly I do not get the error message with the old version of MfSimpleCapture.
As said in my first message I found that the camera I use supports both Interfaces:
IAMCameraControl and IAMVideoControl
I know that not all camera control features are supported. But the one I am looking for: focus is supported.
Could this be the reason?

Excuse me that I bother you again ...

Best regards
Harald Redelberger

**Reply by FactoryXCode** [reply]
Hi Harald, maybe something did go wrong while updating Github. Right now I pushed a one to one copy to GitHub to be shure I tested the sample again with the exact copy from Github you are using and every thing works fine. Maybe you should report the Windows and Delphi versions you are working with. And even so: what camera and softwarerelease (camera drivers) you are using. Because I also tested the drivers on an old Logitech camera (2007), wich has no hardware support at all and works fine allthough without any support for hardware or video controls. So, if your camera is not designed working with Windows 10 or 11, you are in trouble. My HP camera designed to work with Windows 10 works on all settings, ecxcept for some camera setting like focus etc. exposure work fine, but that would be alll.

Regards Tony.

**Reply by HR1959** [reply]
Dear Tony,

First of all thank you again for supporting me.
I would never be able to do what you have done!

Now, I am a big step forward for my capabilities. But step by step:
1.	I tried the sample 2 on a computer with an old external Logilink Web Cam.
And it worked fine as far it can be said because NO camera settings are supported but some video settings are supported and I was able to change them.
2.	Then I connected again my industry camera iDS UV-36L0XC-C and got the error message I reported to you. Then I repeated it with the exe file and not from the IDE. I got the following error message:
![internal error message](https://github.com/user-attachments/assets/510f8643-ef9f-4c75-8933-8d66fa6f512d)
But after pressing ok the camera pictured showed up and I was able to press the now enabled button “Properties”.
The I saw the Camera Settings and the Video settings.
As I reported the camera support focus setting … and it worked!
All other settings are not supported.
And also ALL Video settings are not supported.
My conclusion out of those results AND your last email is that (of course) depending on what the individual camera supports you can control more or less of the settings.
Then I debugged a bit the code and found that the line 1645 “dsVideoControl := m_pSource as IAMVideoProcAmp;” raises the exception and causes the error message. It seems that the camera is not supporting the interface IAMVideoProcAmp at all.
Then I gave a deeper look with the tool I mentioned in my first message: DirectShowCaptureCapabilities-x64.exe.
And you know what? The Camera does really not support the interface IAMVideoProcAmp AT ALL.

The other Logilink camera is reported as supporting both IAMVideoProcAmp and IAMCameraControl. But in reality for the last NO features work, means no features are supported in reality.

So, my final conclusion is that your software does what it should and can ... and my camera is the issue.

Now for learning more about your code I added two trackbars to adjust the settings.
If you like I can send you the changed file. In this way it is more convenient to see setting changes in the video easier.

You asked for the windows version and the Delphi version. I think there is no conjunction with that.
Nevertheless I use windows 10 and 11, both in the latest version and update and Delphi 12.2 meanwhile.

Again thank you very much for your support.

I will now start to integrate your code into my application.

Best regards Harald

**Reply by FactoryXCode** [reply]
Hi Harald,

I'm glad to help you out.
For the code `dsVideoControl := m_pSource as IAMVideoProcAmp`,
 that could raise an exception without a possibility to do a clean response. You could also use `hr := m_pSource.QueryInterface(IID_IAMCameraControl,
                                   dsCameraControl);`

As at this point you can write your own handling of the result of hr, without raised Delphi errors.

If you want a code suggestion, just send a [Pull request.](https://github.com/FactoryXCode/MfPack/pulls).

Regards, Tony

**Reply by HR1959** [reply]
Dear Tony,

I have now integrated all the needed functions I was looking for into my application by generating a simple unit with functions and procedures to handle the focus control in a simple way so that I understand my code also next tomorrow:).

I started my video activities with the help of this source: https://www.delphibasics.info/home/delphibasicsprojects/directxdelphiwebcamcaptureexample

What I want (or need) to do is the following:
1. get the video stream from two industrial camaras uEye UVC XC (see https://de.ids-imaging.com/store/ueye-xc-starter-set-uvc.html) into a bitmap always when a new frame is available
2. reduce the bitmap to the interesting rectangle adjusted and stored before in a setup procedure
3. do some image processing like filtering and edge detection
4. calculate the input for a motion control based on the detected edge position
5. for debugging show both the original interesting rectangle and the edge picture on two individual windows

All this works so far. The annoying issue was that the camera has auto focus. No I am able because of your great support to use either autofocus or manual focus controlled by my software and not by hand in some setup windows.

If I understood properly your work uses the modern version (windows media foundation) and not the old direct show stuff.

Now it would be nice to use your work also for the above mentioned tasks.

But I found "only" solutions which get the video and throw it to a window or into a file.
What is missing is quasi a filter block in between of the where I can to the image processing.

Because your work is huge and I first do not understand everything what you have done and for sure have not seen all possible functions maybe you can give me a hint how to achieve what i explained above with your work as a base.
The main important what I miss is an event somehow called OnNewVideoFrame where I do the image processing.
Can you help me again?
